src/main/java/com/cloudant/client/api/Database.java

```diff
@@ -1029,135 +1029,116 @@ public void ensureFullCommit() {
         try {
-            client.getGson().fromJson(selectorJson, JsonObject.class);
+            selectorObject = getGson().fromJson(selectorJson, JsonObject.class);
         } catch (JsonParseException e) {
             isObject = false;
         }
 
-        if (!isObject) {
-            // needs to start with selector
-            if (!(selectorJson.trim().startsWith("\"selector\""))) {
+        if(!isObject) {
+            if(selectorJson.startsWith("\"selector\"")) {
+                selectorJson = selectorJson.substring(selectorJson.indexOf(":") + 1,
+                        selectorJson.length()).trim();
+                selectorObject = getGson().fromJson(selectorJson, JsonObject.class);
+            } else {
                 throw new JsonParseException("selectorJson should be valid json or like " +
                         "\"selector\": {...} ");
             }
         }
 
-        StringBuilder finalbody = new StringBuilder();
-        if (isObject) {
-            finalbody.append("{\"selector\": ")
-                    .append(selectorJson);
+        if(selectorObject.has("selector")) {
+            indexObject.add("selector", selectorObject.get("selector"));
         } else {
-            //old support
-            finalbody.append("{" + selectorJson);
+            indexObject.add("selector", selectorObject);
         }
 
-        if (rf != null) {
-            finalbody.append(",")
-                    .append(rf.toString());
+        if(fieldsArray.size() > 0) {
+            indexObject.add("fields", fieldsArray);
         }
-        if (so != null) {
-            finalbody.append(",")
-                    .append(so.toString());
+        if(sortArray.size() > 0) {
+            indexObject.add("sort", sortArray);
         }
         if (options.getLimit() != null) {
-            finalbody.append(",")
-                    .append("\"limit\": ")
-                    .append(options.getLimit());
+            indexObject.addProperty("limit", options.getLimit());
         }
         if (options.getSkip() != null) {
-            finalbody.append(",")
-                    .append("\"skip\": ")
-                    .append(options.getSkip());
+            indexObject.addProperty("skip", options.getSkip());
         }
         if (options.getReadQuorum() != null) {
-            finalbody.append(",")
-                    .append("\"r\": ")
-                    .append(options.getReadQuorum());
+            indexObject.addProperty("r", options.getReadQuorum());
         }
         if (options.getUseIndex() != null) {
-            finalbody.append(",")
-                    .append("\"use_index\": ")
-                    .append(options.getUseIndex());
+            JsonArray useIndexArray = new JsonArray();
+            JsonPrimitive element = new JsonPrimitive(options.getUseIndex());
+            useIndexArray.add(element);
+            indexObject.add("use_index", getGson().fromJson(options.getUseIndex(), JsonElement.class));
```

Comment:
Looks like I would need to use a `public` access modifier to call an internal `getGsonUseIndex()` method in `Search`.   How about just changing the return type for `FindByIndexOptions.getUseIndex()` to `JsonArray`? I know it's very possible that it would break code, but we could just add a comment that `toString` can be used?